src/connection/mqtt/mod.rs

```diff
@@ -0,0 +1,844 @@
+use once_cell::sync::OnceCell;
+use rumqttc::{Event as MqttEvent, Packet};
+use serde::{Deserialize, Serialize};
+use tokio::sync::{Mutex, RwLock};
+
+#[cfg(test)]
+pub(crate) use crate::mock::{MockAsyncClient as AsyncClient, MockEventLoop as EventLoop};
+use crate::{
+    builder::{ConnectionConfig, DeviceBuilder, DEFAULT_CHANNEL_SIZE},
+    interface::{
+        mapping::path::MappingPath,
+        reference::{MappingRef, ObjectRef},
+        InterfaceError, Ownership,
+    },
+    interfaces::Interfaces,
+    pairing, properties,
+    retry::DelayedPoll,
+    shared::SharedDevice,
+    store::{error::StoreError, wrapper::StoreWrapper, PropertyStore, StoredProp},
+    topic::ParsedTopic,
+    types::AstarteType,
+    validate::{ValidatedIndividual, ValidatedObject},
+    Interface, Timestamp,
+};
+
+#[cfg(not(test))]
+pub(crate) use rumqttc::{AsyncClient, EventLoop};
+
+use payload::Payload;
+
+use super::{Connection, ReceivedEvent};
+
+/// Borrowing wrapper for the client id
+///
+/// To avoid directly allocating and returning a [`String`] each time
+/// the client id is needed this trait implements [`Display`]
+/// while only borrowing the field needed to construct the client id.
+#[derive(Debug, Clone, Copy)]
+struct ClientId<'a> {
+    realm: &'a str,
+    device_id: &'a str,
+}
+
+impl<'a> Display for ClientId<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}/{}", self.realm, self.device_id)
+    }
+}
+
+/// Shared data of the connection, this struct is internal to the [`Mqtt`] connection
+/// where is wrapped in an arc to share an immutable reference across tasks.
+pub struct SharedMqtt {
+    realm: String,
+    device_id: String,
+    eventloop: Mutex<EventLoop>,
+}
+
+#[derive(Clone)]
+pub struct Mqtt {
+    shared: Arc<SharedMqtt>,
+    client: AsyncClient,
+}
+
+impl Mqtt {
+    /// Initializes values for this struct
+    ///
+    /// This method should only be used for testing purposes since it does not fully
+    /// connect to the mqtt broker as described by the astarte protocol.
+    /// This struct should be constructed with the [`Mqtt::connected`] associated function.
+    fn new(realm: String, device_id: String, eventloop: EventLoop, client: AsyncClient) -> Self {
+        Self {
+            shared: Arc::new(SharedMqtt {
+                realm,
+                device_id,
+                eventloop: Mutex::new(eventloop),
+            }),
+            client,
+        }
+    }
+
+    /// Initializes this connection with the passed settings
+    ///
+    /// The session parameter holds data that will be sent during the
+    /// connection to the astarte server.
+    pub(crate) async fn connected(
+        realm: String,
+        device_id: String,
+        eventloop: EventLoop,
+        client: AsyncClient,
+        session: SessionData,
+    ) -> Result<Self, crate::Error> {
+        let mqtt = Self::new(realm, device_id, eventloop, client);
```

Comment:
I would move the creation of mqtt object outside this method and rename this method into `connect` or `wait_connack` since the `new` method for an Mqtt is private and it is called by the builder.

Reply:
From the previous iterations I tested out, I think this is the best one. `new` isn't called by the builder it's meant to be called from the inside.
The purpose of the `connected` method is to build a connection that is already working; to do it, you need to wait for the `connack` before anything else that's why I perform object creation and connection in a single method. To avoid that someone in the future unwillingly breaks the code by creating an instance of the struct not connected correctly to astarte.

Reply:
I get your point. Still, maybe `connected` isn't a good name since the connection is established within this method.

Reply:
Do you have any idea for alternative names ? I don't know if i can come up with something better tbh.

Reply:
> I think we should use the previous name `wait_for_connack` for connected method, but without arguments, we can assume that if we have an EventLoop and Client, we can be sure that we are making a valid Mqtt struct, so change `new` to `pub (crate) fn new`.

I second this, and I think is the same suggestion @rgallor is making. 

Since the `Mqtt:new` is private, we force the creating of the Mqtt struct through the builder `MqttConfig`, so we can be sure the `Mqtt:wait_for_connack`(without arguments)  is called in the `MqttConfig::connect` method. 

It shouldn't be necessary to make the `Mqtt::new` method pub for the above reason.

Reply:
In the end I left the `new` method as is but renamed the `connected` into `wait_for_connack` as suggested.
This new method which does not build the struct anymore should be called after the `new` to initialize the connection and it expects a `SessionData` struct that holds the introspection, the device properties and server interfaces to subscribe to.